README.md
# Gupload

An easy file uploading solution for simple uploads needs

## Usage

- `[GET] /:filename`
  - Query: If you only pass one of those parameters, the original image will be sent
    - `h`: number // Height of the image
    - `w`: number // Width of the image
- `[POST] /upload`
  - Body: The body needs to have a field named File and a content-type `multipart/form-data;`
    - `file`: File;

## Example

```bash
# Upload

$ curl --location --request POST 'http://0.0.0.0:4242/upload' --form 'file=@/path/to/file'
# Response:
# {
#    "fieldname": "file",
#    "filename": "058c03293925ead7a559216a877d1bc1.jpg",
#    "mimetype": "image/jpeg",
#    "originalname": "pexels-pixabay-36717.jpg",
#    "size": 152231
# }
```

```bash
# Fetch

$ curl --location --request GET 'http://0.0.0.0:4242/058c03293925ead7a559216a877d1bc1.jpg'
# Response:
# The image!

$ curl --location --request GET 'http://0.0.0.0:4242/058c03293925ead7a559216a877d1bc1.jpg?w=400&h=400'
# Response:
# The image in 400x400 (granted we allowed this in the conf)!
```

## Running

## From scratch:

```bash
$ git clone git@github.com:amoutonbrady/gupload.git`
$ pnpm install # (might use yarn or npm if you will, but it was tested only with pnpm)
$ pnpm build # compile TS to JS
$ pnpm start # start the compiled server
$ pnpm dev # do the two commands above in one command
```

## Dockerfile:

`$ docker run -p 4242:4242 -v ./.images:/var/www/images amoutonbrady/gupload`

## Docker-compose:

```yaml
version: "3.8"

services:
  img:
    image: amoutonbrady/gupload
    export:
    
    environments:
      PORT: 4242
      HOST: 0.0.0.0
      MAX_SIZE_MB: "20"
      VALID_SIZES: "['400']"
      ALLOWED_ORIGINS: '["http://0.0.0.0:4242/"]'
    ports:
      - "4242:4242"
    volumes:
      - ./.images:/var/www/images
```

## Configuration & defaults

Configration is done through env variables.

| Variable          | Description                          | Type       | Default                  |
| ----------------- | ------------------------------------ | ---------- | ------------------------ |
| `MAX_SIZE_MB`     | Maximum size per file in MB          | `number`   | 20                       |
| `VALID_SIZES`     | List of sizes available for resizing | `number[]` | [400]                    |
| `ALLOWED_ORIGINS` | List of allowed origins for CORS     | `string[]` | ["http://0.0.0.0:4242/"] |
| `PORT`            | -                                    | `number`   | 4242                     |
| `HOST`            | -                                    | `string`   | "0.0.0.0"                |

## Technologies used

- [Fastify](https://www.fastify.io/): Fast HTTP server with a wide ecosystem of plugins
- [Sharp](https://sharp.pixelplumbing.com/): Image manipulation library (used to resize images)
- [Hasha](https://github.com/sindresorhus/hasha): md5 generation helper


CHANGELOG.md
# Changelog

<a name="0.0.3"></a>
## 0.0.3 (2020-08-22)

### Added

- ✨ Adding changelog [[28b6973](https://github.com/amoutonbrady/gupload/commit/28b697347d4673876e34fdbeeebe66f69a1eefe0)]
- 🎉 initial work [[e20cd6f](https://github.com/amoutonbrady/gupload/commit/e20cd6fd921b2264078f46f6b1d19f55e1b1386b)]

### Fixed

- 🐛 Fix size calculation [[8f09e61](https://github.com/amoutonbrady/gupload/commit/8f09e61966b0da310211480cf477b48d06b989a4)]
- 🐛 Fix bug with host &amp; port not set crashing the Docker [[0b46653](https://github.com/amoutonbrady/gupload/commit/0b466530213b8ac58d341a3141e8a674357fcc46)]

### Miscellaneous

-  Merge branch &#x27;release/0.0.2&#x27; into develop [[b4ced8d](https://github.com/amoutonbrady/gupload/commit/b4ced8d6d0d1cbb8fd6c6311a43b107642d95138)]
- 📝 Fix typos in README [[a0577c3](https://github.com/amoutonbrady/gupload/commit/a0577c3bd9ce035dfdf83542d33b14e8f935c6f3)]
- 📝 Update README [[6499ce3](https://github.com/amoutonbrady/gupload/commit/6499ce37da8b8aa909f41e6314bffd917853dee3)]
-  Merge branch &#x27;release/0.0.1&#x27; into develop [[50bdb65](https://github.com/amoutonbrady/gupload/commit/50bdb65d083936073f5f28e77533b040fce74d04)]




src/resize.ts
import { join, dirname, parse } from "path";
import { preHandlerHookHandler } from "fastify";
import { pathExists, readFile } from "fs-extra";
import sharp from "sharp";

export const makeResizeFileName = (
  file: string,
  height?: string,
  width?: string
) => {
  if (!height && !width) return file;

  const { name, ext } = parse(file);
  const filename = `${name}-${height}x${width}${ext}`;

  return filename;
};

export const resize: (validSizes: number[]) => preHandlerHookHandler = (
  validSizes
) => async (req, rep, done) => {
  const query = req.query as { h?: string; w?: string };
  const params = req.params as { filename: string };
  const dest = join(dirname(__dirname), "images");
  const originalDest = join(dest, params.filename);

  if (query.h && query.w) {
    if (validSizes.includes(+query.h) || validSizes.includes(+query.w)) {
      done();
      return;
    }

    const filename = makeResizeFileName(params.filename, query.h, query.w);
    const finalDest = join(dest, filename);

    const alreadyExists = await pathExists(finalDest);
    if (alreadyExists) return done();

    const originalFile = await readFile(originalDest);

    sharp(originalFile)
      .resize(+query.w, +query.h)
      .toFile(finalDest)
      .then(() => done());
  } else done();
};


src/createUploader.ts
import hasha from "hasha";
import { join, extname } from "path";
import { preHandlerHookHandler } from "fastify";
import { pathExistsSync, outputFile } from "fs-extra";

export function createUploader(dest: string) {
  const single: (fieldname: string, config: Config) => preHandlerHookHandler = (
    fieldname,
    config
  ) => async (req, rep, done) => {
    if (!req.isMultipart()) {
      rep.code(400).send(new Error("Request is not multipart"));
      return;
    }

    const body = (req.body as unknown) as Record<string, WTF[]>;
    const file = body[fieldname] && body[fieldname][0];
    const fileSize = file.data.byteLength / 1e6;

    if (!file) {
      rep.code(400).send(new Error("file is not set"));
      return;
    }

    if (!/image\/.*/g.test(file.mimetype)) {
      rep.code(400).send(new Error("file needs to be of mimetype image/*"));
      return;
    }

    if (fileSize > config.MAX_SIZE_MB) {
      rep
        .code(400)
        .send(
          new Error(
            `file size too big. Got ${fileSize}mb, limit is: ${config.MAX_SIZE_MB}mb`
          )
        );
      return;
    }

    const extension = extname(file.filename);
    const md5 = hasha(file.data, { algorithm: "md5" });

    const name = md5 + extension;
    const finalDestination = join(dest, name);
    const alreadyExist = pathExistsSync(finalDestination);

    if (!alreadyExist) await outputFile(finalDestination, file.data);

    req.file = {
      fieldname,
      filename: name,
      mimetype: file.mimetype,
      originalname: file.filename,
      size: file.data.byteLength,
    };

    done();
  };

  return { single };
}

declare module "fastify" {
  interface FastifyRequest {
    file: File;
  }
}

interface File {
  fieldname: string;
  originalname: string;
  mimetype: string;
  filename: string;
  size: number;
}
interface WTF {
  data: Buffer;
  encoding: string;
  filename: string;
  limit: boolean;
  mimetype: string;
}
interface Config {
  MAX_SIZE_MB: typeof Infinity | number;
}


src/server.ts
import "dotenv/config";
import fastify from "fastify";
import cors from "fastify-cors";
import serve from "fastify-static";
import { join, dirname } from "path";
import multipart from "fastify-multipart";

import { parseEnv } from "./parseEnv";
import { createUploader } from "./createUploader";
import { makeResizeFileName, resize } from "./resize";

const { PORT, HOST, MAX_SIZE_MB, ALLOWED_ORIGINS, VALID_SIZES } = parseEnv({
  PORT: 4242,
  HOST: "0.0.0.0",
  MAX_SIZE_MB: 20,
  ALLOWED_ORIGINS: ["*"],
  VALID_SIZES: [400],
});

const dest = join(dirname(__dirname), "images");
const upload = createUploader(dest);

console.log(ALLOWED_ORIGINS);

const app = fastify();
app.register(cors, { origin: ALLOWED_ORIGINS });
app.register(multipart, { addToBody: true });
app.register(serve, { root: dest, serve: false });

app.route({
  method: "POST",
  url: "/upload",
  preHandler: upload.single("file", { MAX_SIZE_MB }),
  handler: async (req) => req.file,
});

app.route<{
  Params: { filename: string };
  Querystring: { h?: string; w?: string };
}>({
  method: "GET",
  url: "/:filename",
  preHandler: resize(VALID_SIZES),
  async handler(req, rep) {
    const filename = makeResizeFileName(
      req.params.filename,
      req.query.h,
      req.query.w
    );

    return rep.sendFile(filename);
  },
});

app
  .listen(PORT, HOST)
  .then(console.log)
  .catch((e) => {
    console.error(e);
    process.exit(1);
  });
